worksheet_B/Mandelbrot.cpp

```diff
@@ -1,56 +1,232 @@
-	//image.save_bmp("mandelbrot.bmp");
-
-	// Wait for the window to be closed
-	while (!display.is_closed())
-	{
-		display.wait();
-	}
-
-	return 0;
-}
+// Mandelbrot.cpp : Defines the entry point for the console application.
+//
+
+#include "stdafx.h"
+const double width = 800.0, height = 800.0;
+// Took the code from the second comment on http://stackoverflow.com/questions/3018313/algorithm-to-convert-rgb-to-hsv-and-hsv-to-rgb-in-range-0-255-for-both
+typedef struct RgbColor
+{
+	unsigned char r;
+	unsigned char g;
+	unsigned char b;
+} RgbColor;
+
+typedef struct HsvColor
+{
+	unsigned char h;
+	unsigned char s;
+	unsigned char v;
+} HsvColor;
+
+RgbColor HsvToRgb(HsvColor hsv)
+{
+	RgbColor rgb;
+	unsigned char region, remainder, p, q, t;
+
+	if (hsv.s == 0)
+	{
+		rgb.r = hsv.v;
+		rgb.g = hsv.v;
+		rgb.b = hsv.v;
+		return rgb;
+	}
+
+	region = hsv.h / 43;
+	remainder = (hsv.h - (region * 43)) * 6;
+
+	p = (hsv.v * (255 - hsv.s)) >> 8;
+	q = (hsv.v * (255 - ((hsv.s * remainder) >> 8))) >> 8;
+	t = (hsv.v * (255 - ((hsv.s * (255 - remainder)) >> 8))) >> 8;
+
+	switch (region)
+	{
+	case 0:
+		rgb.r = hsv.v; rgb.g = t; rgb.b = p;
+		break;
+	case 1:
+		rgb.r = q; rgb.g = hsv.v; rgb.b = p;
+		break;
+	case 2:
+		rgb.r = p; rgb.g = hsv.v; rgb.b = t;
+		break;
+	case 3:
+		rgb.r = p; rgb.g = q; rgb.b = hsv.v;
+		break;
+	case 4:
+		rgb.r = t; rgb.g = p; rgb.b = hsv.v;
+		break;
+	default:
+		rgb.r = hsv.v; rgb.g = p; rgb.b = q;
+		break;
+	}
+
+	return rgb;
+}
+
+HsvColor RgbToHsv(RgbColor rgb)
+{
+	HsvColor hsv;
+	unsigned char rgbMin, rgbMax;
+
+	rgbMin = rgb.r < rgb.g ? (rgb.r < rgb.b ? rgb.r : rgb.b) : (rgb.g < rgb.b ? rgb.g : rgb.b);
+	rgbMax = rgb.r > rgb.g ? (rgb.r > rgb.b ? rgb.r : rgb.b) : (rgb.g > rgb.b ? rgb.g : rgb.b);
+
+	hsv.v = rgbMax;
+	if (hsv.v == 0)
+	{
+		hsv.h = 0;
+		hsv.s = 0;
+		return hsv;
+	}
+
+	hsv.s = 255 * long(rgbMax - rgbMin) / hsv.v;
+	if (hsv.s == 0)
+	{
+		hsv.h = 0;
+		return hsv;
+	}
+
+	if (rgbMax == rgb.r)
+		hsv.h = 0 + 43 * (rgb.g - rgb.b) / (rgbMax - rgbMin);
+	else if (rgbMax == rgb.g)
+		hsv.h = 85 + 43 * (rgb.b - rgb.r) / (rgbMax - rgbMin);
+	else
+		hsv.h = 171 + 43 * (rgb.r - rgb.g) / (rgbMax - rgbMin);
+
+	return hsv;
+}
+// Initialise the image
+CImg<unsigned char> image(width, height, 1, 3, 0);
+
+// Display the image
+CImgDisplay display(image, "Mandelbrot Set");
+
+// Minimum and maximum coordinates for the fractal
+double minX = -2, maxX = 1, minY = -1.5, maxY = 1.5;
+
+// The sum of the X coordinates the result will be 3;
+double maxXSum = maxX + minX * -1;
+
+// The amount to move one pixel on any axis because it is a 3x3 square Divide it by the amount of pixels which is 800
+double onePix = maxXSum / width;
+
+// Calculate the X an Y coordinate of the center of the screen Starting is currentXc=0.5, currentYc=0.0
+double currentXc = minX + width/2 * onePix;
+double currentYc = maxY - width/2 * onePix;
+
+// To have a referance of how many time I have zoomed in
+int zoomLevel = 0;

```

Comment:
A lot of global variables here, which is generally bad practice. For anything larger than this exercise, you'd want to wrap your variables and functions up into a class.